pkg/engine/turn/turn.go

```diff
@@ -123,67 +140,76 @@ func (mgr *manager) RemoveTarget(id key.TargetID) {
 			break
 		}
 	}
+
 	mgr.order.order = append(mgr.order.order[:idx], mgr.order.order[idx+1:]...)
 }
 
+// StartTurn processes changes to target's gauges on the start of a new turn.
+//	1. Mark target at top of order as "active"
+//	2. get that target's current AV
+//	3. add AV to "TotalAV" (keeps track of how much AV has progressed over entire sim)
+//	4. loop through all targets in the order, subtracing this AV from them (active target gauge = 0)
+//			new_gauge = current_gauge - av * speed
 func (mgr *manager) StartTurn() (key.TargetID, float64, []event.TurnStatus, error) {
 	if mgr.activeTurn {
 		return -1, 0, nil, fmt.Errorf("cannot start turn when already in an active turn: %+v", mgr)
 	}
 
-	// reset gauge cost for this new turn
 	mgr.gaugeCost = 1.0
 	mgr.activeTurn = true
 	mgr.activeTarget = mgr.order.order[0].id
 	av := mgr.av(mgr.activeTarget)
 
-	// mgr.order.order[0].gauge = 0 // set gauge/av of active to 0
-
 	for _, t := range mgr.order.order {
 		t.gauge -= av * mgr.attr.Stats(t.id).SPD()
 	}
```

Comment:
Thinking out loud: I wonder if we need to do a stable sort prior to the logic here. AV is determined by two factors
1. gauge, which the turn manager is responsible for and will re-sort any time it changes
2. speed, which turn manager is **not** responsible for and may change between sorts. In the event that speed changes between the last sort and start of turn, the target at the top of the order may not be the one with the lowest AV.

Reply:
The only thing that appears to change between the last check (in ResetTurn, called in phase2) and beginTurn calling StartTurn are phase2's ActionEnd and ModifierPhase2 modifier ticks. Would those include any SPD-changing effects? I had assumed those would not, since active SPD-buffing skills are called in 'action' to my knowledge, but I suppose that phase2 might include things like SPD buffs expiring?

Even if that's not the case, just to future-proof it, so that even if there aren't any SPD-changing effects now but might be later, I'll go ahead and add a sort.

Reply:
Though I'm wondering: would any of those phase2 effects depend on our gauge reset done in ResetTurn? If not, we could also just call ResetTurn last in phase2, as then nothing would change between phase2 and beginTurn, to my knowledge, so we could be a tiny bit more efficient and not call an extra Sort.

Reply:
>The only thing that appears to change between the last check (in ResetTurn, called in phase2) and beginTurn calling StartTurn are phase2's ActionEnd and ModifierPhase2 modifier ticks

there is also the `executeQueue` call, which will run the logic for any queued up actions and inserts (counters, follow up attacks, ults, extra turns/actions). This ends up being a lot of things. An example of a SPD modifier that happens here would be Asta ult if queued after the active turn action is performed.

>Though I'm wondering: would any of those phase2 effects depend on our gauge reset done in ResetTurn?

Sadly yes, reset needs to happen after the action and prior to/at the start of phase2. Game has odd gauge mechanics based on if you run before or after the reset. A good example is the advance on Sushang ult. It works by setting her gauge to 0 when the ult is performed. In game, if the ult is performed at the start of her turn the advance does nothing/does not happen because her gauge is currently 0. When the reset happens, it will be 10k. If you ult after her action but still on her turn in game, that will cause the advance to work and will result in the next turn also being Sushang. This is because the reset happens immediately after the turn action is performed rather than at the end of the turn

Reply:
Ok, that's interesting; I assumed queued ults would just be coded as new turns. In light of the reset stuff, I'll go ahead and add another sort to StartTurn.